src/include/OSL/rendererservices.h

```diff
@@ -29,6 +30,298 @@ typedef const void* TransformationPtr;
+            return m_type == Type::UInt32;
+        if (std::is_same<T, uint64_t>::value)
+            return m_type == Type::UInt64;
+        if (std::is_same<T, float>::value)
+            return m_type == Type::Float;
+        if (std::is_same<T, double>::value)
+            return m_type == Type::Double;
+        if (std::is_same<T, void*>::value)
+            return m_type == Type::Pointer;
+        if (std::is_same<T, ustring>::value)
+            return m_type == Type::UString;
+        if (std::is_same<T, ustringhash>::value)
+            return m_type == Type::UStringHash;
+
+        return false;
+    }
+
+    template<typename T> T get() const
+    {
+        OSL_DASSERT(false);
+        return T();
+    }
+};
+
+
+template<>
+inline AttributeSpecBuiltinArg
+AttributeSpecArg::get() const
+{
+    OSL_DASSERT(is_holding<AttributeSpecBuiltinArg>());
+    return m_builtin;
+}
+
+template<>
+inline bool
+AttributeSpecArg::get() const
+{
+    OSL_DASSERT(is_holding<bool>());
+    return m_bool;
+}
+
+template<>
+inline int8_t
+AttributeSpecArg::get() const
+{
+    OSL_DASSERT(is_holding<int8_t>());
+    return m_int8;
+}
+
+template<>
+inline int16_t
+AttributeSpecArg::get() const
+{
+    OSL_DASSERT(is_holding<int16_t>());
+    return m_int16;
+}
+
+template<>
+inline int32_t
+AttributeSpecArg::get() const
+{
+    OSL_DASSERT(is_holding<int32_t>());
+    return m_int32;
+}
+
+template<>
+inline int64_t
+AttributeSpecArg::get() const
+{
+    OSL_DASSERT(is_holding<int64_t>());
+    return m_int64;
+}
+
+template<>
+inline uint8_t
+AttributeSpecArg::get() const
+{
+    OSL_DASSERT(is_holding<uint8_t>());
+    return m_uint8;
+}
+
+template<>
+inline uint16_t
+AttributeSpecArg::get() const
+{
+    OSL_DASSERT(is_holding<uint16_t>());
+    return m_uint16;
+}
+
+template<>
+inline uint32_t
+AttributeSpecArg::get() const
+{
+    OSL_DASSERT(is_holding<uint32_t>());
+    return m_uint32;
+}
+
+template<>
+inline uint64_t
+AttributeSpecArg::get() const
+{
+    OSL_DASSERT(is_holding<uint64_t>());
+    return m_uint64;
+}
+
+template<>
+inline float
+AttributeSpecArg::get() const
+{
+    OSL_DASSERT(is_holding<float>());
+    return m_float;
+}
+
+template<>
+inline double
+AttributeSpecArg::get() const
+{
+    OSL_DASSERT(is_holding<double>());
+    return m_double;
+}
+
+template<>
+inline void*
+AttributeSpecArg::get() const
+{
+    OSL_DASSERT(is_holding<void*>());
+    return m_ptr;
+}
+
+template<>
+inline ustring
+AttributeSpecArg::get() const
+{
+    OSL_DASSERT(is_holding<ustring>());
+    return m_ustring;
+}
+
+template<>
+inline ustringhash
+AttributeSpecArg::get() const
+{
+    OSL_DASSERT(is_holding<ustringhash>());
+    return m_ustringhash;
+}
+
+// The AttributeGetterSpec is never in device code, so it can be a bit more
+// complex and have dynamic allocations.  Although we do want to
+// use an interface and not allow direct data member access
+class AttributeGetterSpec {
+    ustring m_function_name;
```

Comment:
Does m_function_name need to be a ustring?  I think its just used as a c_str() later.

Reply:
When we discussed this I agreed that a `ustring` might be too much and a `std::string` might be sufficient. Just rethinking this, I can now see a good use case for `ustring`, it can avoid the many small memory allocations (one per attribute) needed to copy the function name into this struct, since `AttributeGetterSpec::set()` needs to take a member copy.

This can be beneficial when the user has a small number of known function names they could declare statically, e.g.
```
void RendererServices::build_attribute_getter(OSL::ShaderGroup& group, bool object_lookup, OSL::ustring* object_name, OSL::ustring* attribute_name, bool array_lookup, int* array_index, OIIO::TypeDesc type, bool derivatives, OSL::AttributeGetterSpec& spec)
{
  static const OIIO::ustring u_get_global_attribute("rs_get_global_attribute");
  static const OIIO::ustring u_get_object_attribute("rs_get_object_attribute");

  if (object_lookup)
    spec.set(u_get_object_attribute, ...);
  else
    spec.set(u_get_global_attribute, ...);
}
```

I'd be inclined to leave it as a `ustring` now with this in mind?

Reply:
I think the counter case might be the overhead of creating that ustring, it involves a map lookup then insertion into table and copy of string into it.  Which is more than just the std::string copy constructor.

In any event, I think this only happens during the compilation phase and will be dwarfed (performance wise) by most everything else going on.

But I think you might be right here, it is reasonable to expect getters for multiple attributes to share the same function name.  Maybe just add a comment on
`ustring m_function_name; // ustring to mimimize copies when same function used for multiple attributes`